### PR TITLE
QField 💓 its community 

### PR DESCRIPTION
**Pull request by nirvn**
The community spoke, and QField heard you. This PR brings back collapsibility to QField's shiny revamped legend. First, the obligatory GIF:
Peek 2020-09-16 16-28

The legend now respect collapsed state of items when loading a project, as well as when switching map themes. In addition, the item properties popup now has an additional button that allows users to collapse or expand legend items at will.

Kudos for OPENGIS.ch for prioritizing this issue and, as @m-kuhn said earlier, thanks to all the people and organisations who care for QField and bought [a support contract with the sustainability initiative](https://www.opengis.ch/qgis-support/) or committed to a [recurring sponsorship](https://github.com/sponsors/opengisch).

### CODE REVIEW

**Comment by nirvn** [comment]
Note: the expanded / collapsed icons might be oversized, checking the sizing issues as soon as APKs emerge.

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-f422fd719348bf3d013d08205e2b4cabc9dbee2e-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-f422fd719348bf3d013d08205e2b4cabc9dbee2e-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-f422fd719348bf3d013d08205e2b4cabc9dbee2e-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-f422fd719348bf3d013d08205e2b4cabc9dbee2e-x86_64.apk)

**Comment by nirvn** [comment]
Icons properly sized on android hidpi screen device.

So, one UI/UX question to resolve here: at the moment, both a long press and double clicking on an item opens the layer/group item properties popup. We could take on of those and re-assign it to item collapse / expansion (say double clicking). However, if my memory is not tricking me, the duplication was to help out stylus users. @3nids , is that correct? Is it still the case that stylus long press doesn't open the layer item properties popup?

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-f734779006e269565a8ce82cd7e03b9151c17208-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-f734779006e269565a8ce82cd7e03b9151c17208-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-f734779006e269565a8ce82cd7e03b9151c17208-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-f734779006e269565a8ce82cd7e03b9151c17208-x86_64.apk)

**Comment by 3nids** [comment]
>  Is it still the case that stylus long press doesn't open the layer item properties popup?

I don't have it with me now, I'd need to check.

**Comment by signedav** [comment]
Looks nice. What do you think about using a checkbox instead of a combobox, like I described here https://github.com/opengisch/QField/issues/1234#issuecomment-691952098 ?

I think, "Zoom to Layer" and "Tracking" are actions started via this dialog.
And "Show on Canvas" and "Expand" are settings.

It would be great to have a checkbox for both and as soon as you change it, the dialog closes. So it would not need one click more.

**Comment by nirvn** [comment]
I thought about a [x] collapse legend children checkbox option, and ultimately though it'd be better served as a button, with a visual separator. Zoom to layer, tracking, and show on canvas all have effects on the canvas, whereas this new collapse/expand button doesn't, hence why I wanted it a bit separated.

**Comment by nirvn** [comment]
Regarding the show on canvas checkbox, we can easily have it so when someone toggles that checkbox the popup closes.

**Comment by m-kuhn** [comment]
I also think a checkbox would be good.
Or a switch like in the global settings (with support for comments below a title etc)

**Comment by nirvn** [comment]
@m-kuhn , regarding the checkbox, do we want it below the other checkbox, or at the location where the button is ATM?

**Comment by m-kuhn** [comment]
Below sounds good to me. No strong opinion.

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-b3c6e2b61738c1c2236ea93be9a75aa2a7a71b1f-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-b3c6e2b61738c1c2236ea93be9a75aa2a7a71b1f-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-b3c6e2b61738c1c2236ea93be9a75aa2a7a71b1f-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-b3c6e2b61738c1c2236ea93be9a75aa2a7a71b1f-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-78883bc73aeef1bd7adac71a113c2c1f876afd02-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-78883bc73aeef1bd7adac71a113c2c1f876afd02-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-78883bc73aeef1bd7adac71a113c2c1f876afd02-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-78883bc73aeef1bd7adac71a113c2c1f876afd02-x86_64.apk)

**Comment by signedav** [comment]
The property looks super-good. Thanks a lot.

One thing regarding the arrows. I liked the new legend having no arrows. But I see that there need to be an indication that the legend item is expandable. 
On the old legend we used the arrow as the collapse / expand button as well, but since we do not need it as this anymore it can be much less present.
Can the arrows be smaller. Or is there another way to indicate the possibility for expand an item (for the possibility to collapse we wouldn't need one)?

**Comment by nirvn** [comment]
@signedav , I had the same feeling as you do. We can't skip a visual queue, it's too awkward. I first tried to overlay a badge over the legend icons when collapsed but that was super messy. So ended up with what you see, which I've grown to like when field testing this am.

I could try to reduce the width a bit, but that's about all we can do here.

**Comment by signedav** [comment]
Another thought is having a arrow down on the right? So we do not have all items intended even with no kids...
image
But no strong opinion about this, just an input.

Anyway if we keep the current indicators, then I would prefer them smaller and maybe gray (because gray looks for me more like an indicator than something I can click on (like the green does more))...

**Comment by nirvn** [comment]
It's hard to use right edge as we keep it for indicators.

The items align when they are at the same level (ie when an item doesn't have children, there's a space to fill the collapsed/expanded state icon)

I'll give gray a try.

**Comment by signedav** [comment]
Btw. since the last commit https://github.com/opengisch/QField/pull/1295/commits/7cd89195625c3418a2cb7abe39c09ef89aeace1c the tracking-button doesn't work anymore. I don't see the change in the code that made it. If you can't reproduce, then tell me. Then I can checkout and have a look...

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-ba43b16a2dcab13a11bc099a8ba6a9626bc2aec6-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-ba43b16a2dcab13a11bc099a8ba6a9626bc2aec6-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-ba43b16a2dcab13a11bc099a8ba6a9626bc2aec6-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-ba43b16a2dcab13a11bc099a8ba6a9626bc2aec6-x86_64.apk)

File: src/qml/LayerTreeItemProperties.qml [hunk 1]
```diff
@@ -14,28 +14,33 @@ Popup {
...
   padding: 0
 
   onIndexChanged: {
-    itemVisible = layerTree.data(index, FlatLayerTreeModel.Visible)
     title = layerTree.data(index, Qt.DisplayName)
+
+    itemVisibleCheckBox.checked = layerTree.data(index, FlatLayerTreeModel.Visible);
+
+    expandCheckBoxVisible = layerTree.data(index, FlatLayerTreeModel.HasChildren)
+    expandCheckBox.text = layerTree.data( index, FlatLayerTreeModel.Type ) === 'group' ? qsTr('Expand group') : qsTr('Show legend')
```
Comment (signedav) [inline src/qml/LayerTreeItemProperties.qml:35]:
```suggestion
    expandCheckBox.text = layerTree.data( index, FlatLayerTreeModel.Type ) === 'group' ? qsTr('Expand group') : qsTr('Show layer legend')
```
I know, that we decided this in hive. Still while using it it seems strange. You make a double tip an a legend item and there is the option "Show legend" - I can imagine that users think it's the whole legend we are talking about.

**Review by signedav — APPROVED** [review]
Beside the "Show Layer Legend" input, it's very nice and the code looks good.

For me okay to merge. Okay for you as well @m-kuhn ?

**Comment by signedav** [comment]
Screenshot_20200917-115044_QField Beta 1295
Screenshot_20200917-115021_QField Beta 1295

File: src/qml/LayerTreeItemProperties.qml (hunk 1 above)
Comment (nirvn) [reply src/qml/LayerTreeItemProperties.qml:35]: Rule based renderer can create a tree with subfolders of its own. Show layer legend would be odd for these items.

What about show legend item children?

Comment (signedav) [reply src/qml/LayerTreeItemProperties.qml:35]: In the chat yesterday we decided against the children. But I don't mind that much.
Or maybe keep it simple with "Expand legend item"?

Comment (nirvn) [reply src/qml/LayerTreeItemProperties.qml:35]: +1 to expand legend item

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-38c512e8dc1ec35eec54e50629889dc8a846acfe-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-38c512e8dc1ec35eec54e50629889dc8a846acfe-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-38c512e8dc1ec35eec54e50629889dc8a846acfe-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-38c512e8dc1ec35eec54e50629889dc8a846acfe-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-d34f5f875adfe7f5a9c388f6162d25a40e5b4813-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-d34f5f875adfe7f5a9c388f6162d25a40e5b4813-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-d34f5f875adfe7f5a9c388f6162d25a40e5b4813-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1295-d34f5f875adfe7f5a9c388f6162d25a40e5b4813-x86_64.apk)

**Pull request merged**